File: 0056-merge-intervals/0056-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    intervals.sort((a,b) => a[0] - b[0]);
    const results = [];
    for(let interval of intervals) {
        const prev = results.at(-1);
        if (!results.length || prev[1] < interval[0]) {
            results.push(interval);
        } else {
            prev[1] = Math.max(prev[1], interval[1])
        }
    }
    return results;
};

File: 0056-merge-intervals/0056-merge-intervals.rs
func merge(intervals [][]int) [][]int {
    sort.Slice(intervals, func(i, j int) bool {
        return intervals[i][0] < intervals[j][0]
    })
    
    var res [][]int
    i := 0
    for i < len(intervals) {
        left, right := intervals[i][0], intervals[i][1]
        j := i+1
        for j < len(intervals) && intervals[j][0] <= right {
            right = max(right, intervals[j][1])
            j++
        }
        res = append(res, []int{left, right})
        i = j
    }
    
    return res
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}